src/pip/_internal/index.py

```diff
@@ -833,6 +840,15 @@ def get_page(cls, link, skip_archives=True, session=None):
                       "%s" % exc)
             cls._handle_fail(link, reason, url, meth=logger.info)
         except requests.ConnectionError as exc:
+            # blacklist this host for future packages
+            host = urllib_parse.urlparse(url).netloc
+            blacklisted_hosts.add(host)
+            cls._handle_fail(
+                link,
+                "connection error for %s" % host,
+                url,
+                meth=logger.warning
+            )
```

Comment:
Why 2 calls to _handle_fail?

Reply:
I wanted to notify the user without dumping the exception - it has a different log level. exception would be useful in debug mode. unsure of intent of _handle_fail other than common message prefixes, so avoided adding a random call to logging.